integrations: Create incoming webhook for GitHub Sponsors.

drrosa opened the pull request:
Creates an incoming webhook integration for Github Sponsors. The main use case is getting notifications when new sponsors sign up.

Fixes #18320.
Replaces #24836.
CZO [thread](https://chat.zulip.org/#narrow/stream/127-integrations/topic/integration.20for.20Github.20Sponsors/near/1530312).

**Notes:**
- JSON data  for the fixture copied from https://gist.github.com/bgoonz/11331feafe55e4a77d59989380eca965#sponsorship and #20926.

**Screenshots and screen captures:**
![image](https://github.com/zulip/zulip/assets/2343554/bc01fd23-3cc0-405e-9af1-29e140bbef2e)

- `/integrations/doc/githubsponsors`
![image](https://github.com/zulip/zulip/assets/2343554/f8931391-29fb-440d-8863-ba12909c9665)

- `cancelled`
![cancelled](https://github.com/zulip/zulip/assets/2343554/045e1c55-57be-40b2-bc39-5350139c164e)

- `created`
![created](https://github.com/zulip/zulip/assets/2343554/0a5d3039-4867-4c4f-a887-ffaa52207928)

- `edited`
![edited](https://github.com/zulip/zulip/assets/2343554/9ba5e318-468a-441e-876b-f225018883e3)

- `pending_cancellation`
![pending_cancellation](https://github.com/zulip/zulip/assets/2343554/451049d0-10f5-4fef-93e6-045df35d3e21)

- `pending_tier_change`
![pending_tier_change](https://github.com/zulip/zulip/assets/2343554/45f51a87-dcd4-4910-a382-03d3cd351d8a)

- `tier_changed`
![tier_changed](https://github.com/zulip/zulip/assets/2343554/887fce6e-4674-43be-8c7b-65db7bdddfeb)

<details>
<summary>Self-review checklist</summary>

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability

Communicate decisions, questions, and potential concerns.

- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

drrosa commented:
@zulipbot add "integrations (GitHub bot)" "integrations"

drrosa commented:
@alya This PR is ready for review.
@zulipbot add "documentation review" "maintainer review"

sbansal1999 commented on zerver/webhooks/githubsponsors/tests.py:
```diff
@@ -0,0 +1,79 @@
+        # webhook itself; it only needs to be valid JSON.
+        payload = "{}"
+
+        ignored_events = [
+            "edited",
+        ]
```
Is there any particular reason the `edited` event is being ignored here?

drrosa replied:
The main use case of this integration is monitoring new sign ups and changes in subscriptions, but as noted [here](https://docs.github.com/en/webhooks/webhook-events-and-payloads?actionType=edited#sponsorship) "the edited event types include the details about the change when someone edits a sponsorship to change the privacy". I figured it's probably not relevant for this use case, but maybe we should implement it anyway?

sbansal1999 commented on zerver/webhooks/githubsponsors/view.py:
```diff
@@ -0,0 +1,174 @@
+fixture_to_headers = get_http_headers_from_filename("HTTP_X_GITHUB_EVENT")
+
+
+class Helper:
+    def __init__(
+        self,
+        request: HttpRequest,
+        payload: WildValue,
+        include_title: bool,
+    ) -> None:
+        self.request = request
+        self.payload = payload
+        self.include_title = include_title
+
+
+def get_cancelled_body(helper: Helper) -> str:
+    payload = helper.payload
+    template = "{user_name} cancelled their {subscription} subscription."
+    return template.format(
+        user_name=get_sender_name(payload),
+        subscription=get_subscription(payload),
+    ).rstrip()
```
Probably you can just drop the usage of the `Helper` class here as its properties such as `include_title` are not being used anywhere.

```suggestion
def get_cancelled_body(payload: WildValue) -> str:
    template = "{user_name} cancelled their {subscription} subscription."
    return template.format(
        user_name=get_sender_name(payload),
        subscription=get_subscription(payload),
    ).rstrip()
```

drrosa replied:
Yeah, that's good point. I was using the GitHub integration as template, and thought these properties might come in handy to add more functionality to this integration in the future. Is it bad practice to leave them there for future potential use cases?

sbansal1999 replied:
We can leave it for future potential but I don't think it would take huge refactoring to implement something like this. Right now, it's just taking one more mental step to understand the whole flow and hence isn't useful here.

sbansal1999 reviewed (commented):
Thanks, @drrosa for working on this one.

This mostly looks good to me, just left a couple of comments.

alya commented:
I don't have any feedback here, so I think you can go ahead and mark this for integration review once @sbansal1999 's feedback has been addressed.

drrosa commented:
> I don't have any feedback here, so I think you can go ahead and mark this for integration review once @sbansal1999 's feedback has been addressed.

I updated the PR + screenshots, and addressed @sbansal1999's feedback by handling the `edited` event and dropping the `Helper` class. 
@zulipbot  add "integration review"

sbansal1999 commented:
Thanks for resolving the comments. This looks good to me now.

timabbott commented:
We're discussing if this should be part of the main GitHub integration at the code level (while still maybe being independent in the documentation).

drrosa commented:
@timabbott I reimplemented this as part of the main GitHub integration + independent documentation page.

timabbott commented:
This looks great, merged, thanks @drrosa!
